headers/types/common/common.h

```diff
@@ -36,6 +53,47 @@ struct file_stream {
 };
 ASSERT_SIZE(struct file_stream, 72);
 
+// Struct used to load, store and unload WTE files
+struct wte_handle {
+    void* content; // Pointer to the heap-allocated WTE data. Only stored for freeing the data
+    struct wte_header* header;
+};
+ASSERT_SIZE(struct wte_handle, 8);
+
+// These arguments are almost directly passed to the TEXIMAGE_PARAM register, just rearranged
+// For more information see:
+// https://problemkaputt.de/gbatek.htm#ds3dtextureattributes
+struct wte_texture_params {
+    uint8_t texture_smult : 3;
+    uint8_t texture_tmult : 3;
+    uint8_t unused6 : 2;
+    enum texture_format format : 3;
+    bool flip_x : 1;
+    bool flip_y : 1;
+};
+ASSERT_SIZE(struct wte_texture_params, 2);
+
+// To accurately describe the structure of the header we need to enable packing
+#pragma pack(push, 4)
```

Comment:
There's already a global `#pragma pack(1)` active in all headers, so you shouldn't need to mess with the packing. I think it's also easier to reason about structure if padding is stated explicitly. So as defined, this should be (assuming this is the packing you intended):

```c
struct wte_header {
    char signature[4]; // 0x0: Signature bytes (must be "\x57\x54\x45\x00")
    void* texture; // 0x4
    uint32_t texture_size; // 0x8
    struct wte_texture_params params; // 0xC
    uint16_t _padding_0xe;
    /*  These bounds are NOT used by the game, but they prove useful to extract the texture out
        of the file. The offsets are redundant and should be zero
        The width specified here should always be the same as the one specified in the texture
        params, but the height may be lower. The reason is that just like the width, the height
        needs to be a power of 2 in the range of 8..1024. The actual texture can have a lower
        height, but not a lower width, as the width is required to properly read the image */
    struct rect16_xywh texture_bounds; // 0x10
    struct rgbx* palette; // 0x18
    uint16_t color_amt; // 0x1C: How many colors are stored in the palette
    uint16_t _padding_0x1e;
};
ASSERT_SIZE(struct wte_header, 32);
```

Also, this is a fairly large struct so let's put the field offsets in the comments as above. They're convenient to have when looking in a RAM viewer.

Reply:
About the padding, Yeah I can see how explicit padding would be nicer (but it makes harder to imagine that what you have right there was the struct used in the games almost 15 years ago :P)

I've actually never seen field offsets interspersed in a struct. I wonder if this will always be a requirement for big structs?

Reply:
> (but it makes harder to imagine that what you have right there was the struct used in the games almost 15 years ago :P)

Yeah that's probably true. I think it's more useful for us from a pragmatic reverse engineering standpoint though. When you have the full source code as a dev you don't usually need to worry about raw struct layouts in the first place, but we do :P.

> I've actually never seen field offsets interspersed in a struct. I wonder if this will always be a requirement for big structs?

Well if it's too annoying then you don't have to. Maybe it's not so useful in this case. For stuff like `struct dungeon` it's really nice to have as a reference when live-debugging in an emulator, since you can't really import struct definitions into a memory viewer, and it lets the C types serve the same purpose as, e.g., something like [my RAM notes](https://docs.google.com/document/d/1_Q_7BGmNx5wJtJ9iJEwlK1WITjiCcEQxE9C82RECJbg).

Reply:
> When you have the full source code as a dev you don't usually need to worry about raw struct layouts in the first place, but we do :P.
>
Heh, I can agree with that
> since you can't really import struct definitions into a memory viewer, and it lets the C types serve the same purpose as, e.g., something like [my RAM notes](https://docs.google.com/document/d/1_Q_7BGmNx5wJtJ9iJEwlK1WITjiCcEQxE9C82RECJbg).
>
Actually, there might be a way to use them in an emulator. As you might be aware, Desmume has a Lua scripting engine. With it I've been able to make a small type subsystem so you can bind defined types (including structs!) to addresses and read/write their contents in a typed manner. It's still in a pretty early stage, as I didn't expect to share it with anyone, but maybe we could use this concept here. It'd require parsing the structs (or just add them manually which is what I've been doing), but I'm pretty sure it should be doable (and IMO, it's way more handy than a RAM viewer). I wonder what are your thoughts on this?

Reply:
> I've been able to make a small type subsystem so you can bind defined types (including structs!) to addresses and read/write their contents in a typed manner

That sounds really interesting. I would love to see how that looks and how easy it is to use.

For this repo, I still think field offsets have a lower barrier to entry to use, since it sounds like your method would require a person to set up Lua scripting on DeSmuME (which I believe is only supported on Windows). Minimizing environment setup was something I strived for with this repo, as you can maybe tell from how I set up all this CI stuff so people wouldn't need to set up a dev environment.

But! I would rather have some contribution than none, so if you find labeling field offsets unnecessary or annoying, feel free to just not do it :P. We're not actually losing information so they can always be added later if needed.

Reply:
Yeah, It's a shame that Lua scripting and Desmume in general don't work for Linux (I couldn't get it to work at least), being the best emulator for debugging there is. I'd be glad to share the prototype, but I might need some advice regarding how should I share it...

> Minimizing environment setup was something I strived for with this repo
>
Hah, that's actually a very nice thought, setting environments is a pain isn't it? >_< But in the end, it'd only be a lil' upgrade and wouldn't provide anything more than doing it the old fashioned way (other than some convenience :P)

Reply:
> I'd be glad to share the prototype, but I might need some advice regarding how should I share it...

Is it just a Lua script? I think putting it up on GitHub would be a natural way to share it.

Reply:
Hm alright... I put it on the fork of this repo, I'm not sure if that's what you meant
You should be able to download the files and test it for yourself. It comes with a lil' example, although there's some documentation missing. Also it might be worth it to move this conversation elsewhere :P (where to, I'm not really sure)